app/code/Magento/CacheInvalidate/Observer/InvalidateVarnishObserver.php

```diff
@@ -57,10 +71,21 @@ public function execute(\Magento\Framework\Event\Observer $observer)
         if ($this->config->getType() == \Magento\PageCache\Model\Config::VARNISH && $this->config->isEnabled()) {
             $bareTags = $this->getTagResolver()->getTags($object);
 
+            $tagsBatchSize = 0;
             $tags = [];
-            $pattern = "((^|,)%s(,|$))";
+            $pattern = '((^|,)%s(,|$))';
             foreach ($bareTags as $tag) {
-                $tags[] = sprintf($pattern, $tag);
+                $formattedTag = sprintf($pattern, $tag);
+
+                // Send request if batch size is reached and add the implode with pipe to the computation
+                if ($tagsBatchSize + strlen($formattedTag) > $this->requestSize - count($tags) - 1) {
+                    $this->purgeCache->sendPurgeRequest(implode('|', array_unique($tags)));
```

Comment:
why do we need send `sendPurgeRequest` twice? Line 82 and 91

I understand that in line 91 you will purge leftovers, but I would like to keep request configuration (payloads) and requests itself separately. 

Let's implement a service for batches generation and cover it with the tests.
Use new service in the observer. The observer should simply send purge requests prepared by the new service.

This will allow customize this service for 3rd party developers in case they need some other behavior. 
Make it configurable and customizable.